File: run_single_bayesm_experiment.R
# ==============================================================================
# Run ONE mixture-HBMNL fit with bayesm::rhierMnlRwMixture and dump the raw
# posterior draws in a portable form for the Python side to convert into the
# canonical posterior_raw.pkl format (identical to the Liesel runners).
#
# This script does PURE SAMPLING only. It does not write pickles, meta.json or
# status.json - that is the job of run_single_bayesm_experiment.py, which calls
# this script as a subprocess. Keeping R limited to sampling means the artifact
# format / pickling logic lives in one place (Python), exactly mirroring how
# run_single_experiment.py owns the Liesel artifacts.
#
# Multiple chains are produced via a SEED LOOP (Rossi's convention is a single
# long chain; for the cross-sampler convergence comparison we stack C seed-based
# chains to (chains, draws, ...), matching the Liesel 1/2-chain structure).
#
# Bridge format written into --out-raw-dir:
#     dims.json                  shapes + axis order for every array
#     meta_r.json                sampling metadata (seeds, timings, loglike, ...)
#     mu_chain<c>.bin            float64 LE, R-dim (S, K, P)
#     prec_chain<c>.bin          float64 LE, R-dim (S, K, P, P)   (Sigma^{-1})
#     pvec_chain<c>.bin          float64 LE, R-dim (S, K)
#     beta_chain<c>.bin          float64 LE, R-dim (S, N, P)
#     delta_chain<c>.bin         float64 LE, R-dim (S, D, P)      (only if Z)
#
# Every .bin is written column-major (R's native order); Python reshapes each
# with order="F" using the R dims recorded in dims.json. We export the PRECISION
# (Sigma^{-1} = rooti %*% t(rooti), bayesm's own definition) rather than a
# covariance reconstructed via chol2inv(), which is convention-robust and lets
# Python build the exact TFP FillScaleTriL latent the analysis code expects.
# ==============================================================================

suppressMessages(library(bayesm))
suppressMessages(library(jsonlite))


# -- minimal --key value argument parser --------------------------------------
parse_args <- function(args) {
    out <- list()
    i <- 1
    while (i <= length(args)) {
        key <- sub("^--", "", args[i])
        val <- args[i + 1]
        out[[key]] <- val
        i <- i + 2
    }
    out
}

a <- parse_args(commandArgs(trailingOnly = TRUE))

DATA_PATH   <- a[["data-path"]]
OUT_RAW_DIR <- a[["out-raw-dir"]]
SCENARIO    <- if (!is.null(a[["scenario"]])) a[["scenario"]] else "NA"
N_COMP      <- as.integer(a[["k-model"]])
CHAINS      <- as.integer(a[["chains"]])
R_TOTAL     <- as.integer(a[["r-total"]])
BURN_IN     <- as.integer(a[["burn-in"]])
THIN        <- as.integer(a[["thin"]])
SEED        <- as.integer(a[["seed"]])
A_MU        <- as.numeric(a[["a-mu"]])
A_DELTA     <- as.numeric(a[["a-delta"]])
DIRICHLET_A <- as.numeric(a[["dirichlet-a"]])
NPRINT      <- if (!is.null(a[["nprint"]])) as.integer(a[["nprint"]]) else 2000L

stopifnot(!is.null(DATA_PATH), !is.null(OUT_RAW_DIR),
          !is.na(N_COMP), !is.na(CHAINS), !is.na(R_TOTAL),
          !is.na(BURN_IN), !is.na(THIN), !is.na(SEED))

dir.create(OUT_RAW_DIR, recursive = TRUE, showWarnings = FALSE)

cat("============================================================\n")
cat(sprintf("bayesm rhierMnlRwMixture | scenario=%s\n", SCENARIO))
cat(sprintf("Data file : %s\n", DATA_PATH))
cat(sprintf("K_MODEL=%d  chains=%d  R=%d  thin=%d  burn-in=%d  seed=%d\n",
            N_COMP, CHAINS, R_TOTAL, THIN, BURN_IN, SEED))
cat(sprintf("Amu=%g  Ad=%g  dirichlet_a=%g\n", A_MU, A_DELTA, DIRICHLET_A))
cat("============================================================\n")


# -- 1. load data -------------------------------------------------------------
raw <- fromJSON(DATA_PATH, simplifyVector = TRUE)

n_units  <- as.integer(raw$n_units)
n_params <- as.integer(raw$n_params)
n_alts   <- as.integer(raw$n_alts)
n_demos  <- as.integer(raw$n_demos)
K_true   <- as.integer(raw$K)
n_obs    <- as.integer(length(raw$y) / n_units)

P <- n_params
D <- n_demos
N <- n_units
has_Z <- !is.null(raw$Z) && D > 0

cat(sprintf("N=%d  obs/unit=%d  alts=%d  P=%d  D=%d  K_true=%d\n",
            N, n_obs, n_alts, P, D, K_true))

# raw$X must simplify to a (n_total, n_alts, n_params) array.
if (length(dim(raw$X)) != 3) {
    stop("raw$X did not simplify to a 3D array; check the data JSON.")
}

# -- 2. reconstruct lgtdata (bayesm layout) -----------------------------------
# Rows of X_all are ordered occasion-major (alt fastest), occasions ordered
# unit-major - matching how dgp.py flattens (unit, obs) with an (n_alts,P) block.
X_aperm <- aperm(raw$X, c(2, 1, 3))                 # (n_alts, n_total, P)
X_all   <- matrix(X_aperm, ncol = n_params, byrow = FALSE)
y_all   <- as.integer(unlist(raw$y)) + 1L           # 0-indexed -> 1-indexed

lgtdata <- vector("list", N)
rows_per_unit <- n_obs * n_alts
for (i in seq_len(N)) {
    r0 <- (i - 1L) * rows_per_unit + 1L
    r1 <- i * rows_per_unit
    o0 <- (i - 1L) * n_obs + 1L
    o1 <- i * n_obs
    lgtdata[[i]] <- list(y = y_all[o0:o1], X = X_all[r0:r1, , drop = FALSE])
}

if (has_Z) {
    Z <- matrix(unlist(raw$Z), nrow = N, ncol = D, byrow = FALSE)
    data_list <- list(p = n_alts, lgtdata = lgtdata, Z = Z)
} else {
    data_list <- list(p = n_alts, lgtdata = lgtdata)
}

# -- 3. prior matched to the Liesel mixture model -----------------------------
#   Sigma_k^{-1} ~ W(nu, V^{-1}),  nu = P + 3,  V = nu * I
#   mu_k | Sigma_k ~ N(0, Sigma_k / Amu)
#   vec(Delta)     ~ N(0, Ad^{-1}),  Ad = A_delta * I
#   pvec           ~ Dirichlet(a),   a = dirichlet_a * 1
Prior <- list(
    ncomp = N_COMP,
    nu    = P + 3,
    V     = (P + 3) * diag(P),
    Amu   = A_MU,
    a     = rep(DIRICHLET_A, N_COMP)
)
if (has_Z) Prior$Ad <- A_DELTA * diag(D * P)

# keep = 1: bayesm returns EVERY raw draw. We discard the warmup and thin
# ourselves below, so the burn-in is removed in raw-iteration units BEFORE any
# thinning - rather than thinning first and dropping burn-in in thinned units.
# (Trade-off: the returned chain is unthinned, so betadraw is ~THIN x larger in
# memory; runtime is unaffected since bayesm always runs all R iterations.)
Mcmc <- list(R = R_TOTAL, keep = 1L, nprint = NPRINT)


# -- 4. draw bookkeeping (identical across chains -> stackable) ----------------
# Order of operations: keep all -> discard warmup -> thin, in RAW iteration units.
#   1. bayesm returned all R_TOTAL draws (keep = 1 above).
#   2. drop the first BURN_IN draws (warmup).
#   3. keep every THIN-th of what remains.
# keep_idx therefore holds raw iteration numbers BURN_IN+1, BURN_IN+1+THIN, ...
keep_idx <- seq.int(BURN_IN + 1L, R_TOTAL, by = THIN)
S        <- length(keep_idx)
cat(sprintf("Raw draws/chain=%d  warmup discarded=%d  thin=%d  retained=%d\n",
            R_TOTAL, BURN_IN, THIN, S))

write_bin <- function(arr, path) {
    con <- file(path, "wb")
    writeBin(as.double(arr), con, size = 8, endian = "little")
    close(con)
}


# -- 5. seed loop over chains -------------------------------------------------
seeds        <- integer(CHAINS)
durations_s  <- numeric(CHAINS)
loglike_mean <- numeric(CHAINS)
t_start      <- Sys.time()

for (chain in seq_len(CHAINS)) {
    cidx  <- chain - 1L
    cseed <- SEED + cidx
    seeds[chain] <- cseed
    cat(sprintf("\n--- chain %d / %d  (seed=%d) ---\n", chain, CHAINS, cseed))

    set.seed(cseed)
    c_t0 <- Sys.time()
    out  <- rhierMnlRwMixture(Data = data_list, Prior = Prior, Mcmc = Mcmc)
    durations_s[chain] <- as.numeric(difftime(Sys.time(), c_t0, units = "secs"))

    # mu (S,K,P) and precision (S,K,P,P) from the component draws
    mu_arr   <- array(0.0, dim = c(S, N_COMP, P))
    prec_arr <- array(0.0, dim = c(S, N_COMP, P, P))
    cd <- out$nmix$compdraw
    for (s in seq_len(S)) {
        comp <- cd[[keep_idx[s]]]
        for (k in seq_len(N_COMP)) {
            mu_arr[s, k, ] <- comp[[k]]$mu
            rt <- comp[[k]]$rooti                 # inverse upper-Cholesky of Sigma
            prec_arr[s, k, , ] <- rt %*% t(rt)    # = Sigma^{-1} (bayesm convention)
        }
    }

    # pvec (S,K) - probdraw may be (K, R_draws) or (R_draws, K)
    pd <- out$nmix$probdraw
    if (nrow(pd) == N_COMP) pd <- t(pd)
    pvec_arr <- pd[keep_idx, , drop = FALSE]

    # beta (S,N,P) from (N,P,R_draws)
    beta_arr <- aperm(out$betadraw[, , keep_idx, drop = FALSE], c(3, 1, 2))

    write_bin(mu_arr,   file.path(OUT_RAW_DIR, sprintf("mu_chain%d.bin",   cidx)))
    write_bin(prec_arr, file.path(OUT_RAW_DIR, sprintf("prec_chain%d.bin", cidx)))
    write_bin(pvec_arr, file.path(OUT_RAW_DIR, sprintf("pvec_chain%d.bin", cidx)))
    write_bin(beta_arr, file.path(OUT_RAW_DIR, sprintf("beta_chain%d.bin", cidx)))

    if (has_Z) {
        # Deltadraw row = vec(Delta) column-major of bayesm's (nvar x nz) = (P, D).
        # aperm to (S, D, P) so Python reads it as (C, S, D, P) matching TRUE_DELTA.
        Dd <- out$Deltadraw[keep_idx, , drop = FALSE]
        delta_arr <- aperm(array(t(Dd), dim = c(P, D, S)), c(3, 2, 1))
        write_bin(delta_arr, file.path(OUT_RAW_DIR, sprintf("delta_chain%d.bin", cidx)))
    }

    loglike_mean[chain] <- mean(out$loglike[keep_idx])
    cat(sprintf("chain %d done in %.1fs  mean loglike=%.1f\n",
                chain, durations_s[chain], loglike_mean[chain]))
    rm(out); gc(verbose = FALSE)
}

total_duration_s <- as.numeric(difftime(Sys.time(), t_start, units = "secs"))


# -- 6. write the bridge metadata ---------------------------------------------
per_chain <- list(
    mu   = c(S, N_COMP, P),
    prec = c(S, N_COMP, P, P),
    pvec = c(S, N_COMP),
    beta = c(S, N, P)
)
if (has_Z) per_chain$delta <- c(S, D, P)

dims <- list(
    chains = CHAINS, n_samples = S, K = N_COMP, P = P, D = D, N = N,
    has_Z = has_Z, order = "F", per_chain = per_chain
)
write_json(dims, file.path(OUT_RAW_DIR, "dims.json"), auto_unbox = TRUE)

meta_r <- list(
    scenario = SCENARIO, k_model = N_COMP, k_true = K_true,
    n_params = P, n_units = N, n_demos = D, n_obs = n_obs, n_alts = n_alts,
    chains = CHAINS, r_total = R_TOTAL, thin = THIN, burn_in = BURN_IN,
    n_samples = S, a_mu = A_MU, a_delta = A_DELTA, dirichlet_a = DIRICHLET_A,
    # as.list keeps these as JSON arrays even when CHAINS == 1 (jsonlite's
    # auto_unbox would otherwise collapse a length-1 vector to a scalar).
    seeds = as.list(seeds), durations_s = as.list(round(durations_s, 1)),
    total_duration_s = round(total_duration_s, 1),
    loglike_mean = as.list(round(loglike_mean, 2)),
    bayesm_version = as.character(packageVersion("bayesm")),
    r_version = R.version.string,
    started_at = format(t_start, "%Y-%m-%dT%H:%M:%S")
)
write_json(meta_r, file.path(OUT_RAW_DIR, "meta_r.json"), auto_unbox = TRUE)

cat(sprintf("\nAll chains complete in %.1fs. Raw draws -> %s\n",
            total_duration_s, OUT_RAW_DIR))

File: run_single_bayesm_experiment.py
"""
Run ONE mixture-HBMNL fit with bayesm and persist the SAME artifacts the Liesel
runner does, so the identical analysis notebook + src/analysis.py work unchanged.

This is the bayesm counterpart of run_single_experiment.py. It:
  1. invokes run_single_bayesm_experiment.R as a subprocess (pure sampling),
     which dumps raw float64 draws + dims.json/meta_r.json into <outdir>/_bayesm_raw/,
  2. converts those raw draws into the canonical posterior_raw.pkl dict, with the
     EXACT keys/shapes the Liesel pipeline produces:
         mu_k                     (chains, draws, K, P)
         sigma_inv_chol_k_latent  (chains, draws, K, P(P+1)/2)  - TFP FillScaleTriL
         pvec                     (chains, draws, K)
         Delta                    (chains, draws, D, P)         - only if demographics
         beta_i                   (chains, draws, N, P)
  3. writes export.pkl (via the shared export_posterior_to_pickle), meta.json,
     status.json, summary.txt and sampling.log - same names the notebook expects.

The crucial detail: bayesm samples Sigma_k directly (Gibbs), with no Cholesky-of-
precision latent. We rebuild that latent so the SAME analysis code can invert it
back to Sigma. R exports the precision Sigma^{-1} = rooti @ rooti.T (bayesm's own
definition); here we take its lower Cholesky L (L L.T = precision) and map it
through FillScaleTriL().inverse - the exact representation the Liesel model stores.

Invoked by run_all_experiments.py (--samplers bayesm) as a subprocess (one process per fit).
"""

import argparse
import datetime
import json
import pathlib
import pickle
import shutil
import subprocess
import sys
import time
import traceback

import numpy as np


# --------------------------------------------------------------------------- #
# Project root resolution (same idiom as run_single_experiment.py)
# --------------------------------------------------------------------------- #
PROJECT_ROOT = next(
    p for p in [pathlib.Path(__file__).resolve(), *pathlib.Path(__file__).resolve().parents]
    if (p / "pyproject.toml").exists()
)
if str(PROJECT_ROOT) not in sys.path:
    sys.path.insert(0, str(PROJECT_ROOT))

R_SCRIPT = PROJECT_ROOT / "run_single_bayesm_experiment.R"

# Rscript location: env override wins, else the known Windows install, else PATH.
import os
_DEFAULT_RSCRIPT = r"C:\Program Files\R\R-4.5.1\bin\Rscript.exe"
RSCRIPT = os.environ.get("BAYESM_RSCRIPT") or (
    _DEFAULT_RSCRIPT if pathlib.Path(_DEFAULT_RSCRIPT).exists() else "Rscript"
)


def parse_args():
    ap = argparse.ArgumentParser(description="Run one mixture-HBMNL fit with bayesm.")
    ap.add_argument("--scenario", required=True)
    ap.add_argument("--k-model", type=int, required=True)
    ap.add_argument("--chains", type=int, default=1)
    ap.add_argument("--seed", type=int, default=42)
    ap.add_argument("--outdir", required=True)
    ap.add_argument("--a-delta", type=float, default=0.01)
    ap.add_argument("--a-mu", type=float, default=0.0625)
    ap.add_argument("--dirichlet-a", type=float, default=1.0)
    # bayesm MCMC controls. The R side keeps ALL raw draws, discards the first
    # burn_in (warmup) iterations, THEN thins by `thin` - in that order, in raw
    # iteration units. Retained/chain = len(range(burn_in+1, r_total, thin)):
    #   (42000 - 2000) / 4 = 10000 retained draws, matching the Liesel chains.
    ap.add_argument("--r-total", type=int, default=42000)
    ap.add_argument("--burn-in", type=int, default=2000)
    ap.add_argument("--thin", type=int, default=4)
    ap.add_argument("--nprint", type=int, default=2000)
    ap.add_argument("--data-dir", default=None)
    ap.add_argument("--keep-raw", action="store_true",
                    help="Keep the large _bayesm_raw/*.bin files (default: delete after conversion).")
    return ap.parse_args()


def _read_bin(path, shape):
    """Read an R column-major float64 dump and reshape with Fortran order."""
    arr = np.fromfile(path, dtype="<f8")
    return arr.reshape(tuple(shape), order="F")


def _precision_to_latent(prec):
    """
    prec : (C, S, K, P, P) precision matrices (Sigma^{-1}).
    Returns the TFP FillScaleTriL latent (C, S, K, P(P+1)/2) such that
    recover_covariance_matrices() / _sigma_from_latent() round-trip back to Sigma.
    """
    import jax.numpy as jnp
    import tensorflow_probability.substrates.jax.bijectors as tfb

    # Symmetrise to kill tiny asymmetry, then lower-Cholesky: L L^T = precision.
    prec = 0.5 * (prec + np.swapaxes(prec, -1, -2))
    L = np.linalg.cholesky(prec)                       # lower-triangular, positive diag
    latent = np.asarray(tfb.FillScaleTriL().inverse(jnp.asarray(L)))
    return latent


def main():
    args = parse_args()

    outdir = pathlib.Path(args.outdir)
    outdir.mkdir(parents=True, exist_ok=True)
    raw_dir = outdir / "_bayesm_raw"

    status_path  = outdir / "status.json"
    summary_path = outdir / "summary.txt"
    sampling_log = outdir / "sampling.log"

    data_dir = (pathlib.Path(args.data_dir) if args.data_dir
                else PROJECT_ROOT / "data" / "simulated" / "mixture")
    data_path = data_dir / f"{args.scenario}.json"

    summary_lines = []

    def hlog(msg=""):
        print(msg, flush=True)
        summary_lines.append(msg)

    started_at = datetime.datetime.now().isoformat(timespec="seconds")
    t0 = time.time()

    try:
        if not data_path.exists():
            raise FileNotFoundError(f"Data file not found: {data_path}")

        with open(data_path) as f:
            raw = json.load(f)
        K_TRUE = int(raw["K"])
        P = int(raw["n_params"])
        N = int(raw["n_units"])
        D = int(raw.get("n_demos", 0))

        hlog("=" * 60)
        hlog(f"Scenario           : {args.scenario}")
        hlog(f"Data file          : {data_path}")
        hlog(f"Units (N)          : {N}")
        hlog(f"Parameters (P)     : {P}")
        hlog(f"Demographics (D)   : {D}")
        hlog(f"K in data (K_TRUE) : {K_TRUE}")
        if raw.get("TRUE_PVEC") is not None:
            hlog(f"TRUE_PVEC          : {np.array(raw['TRUE_PVEC']).round(4)}")
        hlog("-" * 60)
        hlog(f"Sampler            : bayesm (rhierMnlRwMixture)")
        hlog(f"K_MODEL            : {args.k_model}")
        hlog(f"dirichlet_a        : {args.dirichlet_a}  | a_mu: {args.a_mu} | A_delta: {args.a_delta}")
        hlog(f"chains/R/thin/burn : {args.chains} / {args.r_total} / {args.thin} / {args.burn_in}")
        hlog(f"Rscript            : {RSCRIPT}")
        hlog("=" * 60)

        # ── 1. run the R sampler ───────────────────────────────────────────
        raw_dir.mkdir(parents=True, exist_ok=True)
        r_cmd = [
            RSCRIPT, str(R_SCRIPT),
            "--data-path",   str(data_path),
            "--out-raw-dir", str(raw_dir),
            "--scenario",    args.scenario,
            "--k-model",     str(args.k_model),
            "--chains",      str(args.chains),
            "--r-total",     str(args.r_total),
            "--burn-in",     str(args.burn_in),
            "--thin",        str(args.thin),
            "--seed",        str(args.seed),
            "--a-mu",        str(args.a_mu),
            "--a-delta",     str(args.a_delta),
            "--dirichlet-a", str(args.dirichlet_a),
            "--nprint",      str(args.nprint),
        ]
        hlog(f"Launching R sampler ...")
        # Stream R's stdout/stderr into sampling.log (the clean engine log).
        with open(sampling_log, "w", encoding="utf-8") as logf:
            proc = subprocess.run(
                r_cmd, stdout=logf, stderr=subprocess.STDOUT,
                cwd=str(PROJECT_ROOT),   # so the project .Rprofile activates renv
            )
        if proc.returncode != 0:
            tail = sampling_log.read_text(encoding="utf-8", errors="replace").splitlines()[-25:]
            raise RuntimeError(
                "R sampler failed (exit "
                f"{proc.returncode}). Last lines of sampling.log:\n" + "\n".join(tail)
            )

        # ── 2. load the raw draws ──────────────────────────────────────────
        with open(raw_dir / "dims.json") as f:
            dims = json.load(f)
        with open(raw_dir / "meta_r.json") as f:
            meta_r = json.load(f)
        # jsonlite auto-unboxes length-1 vectors to scalars, so a chains==1 run
        # yields a bare int where the per-chain list is expected. Force list form.
        def _as_list(x):
            if x is None:
                return []
            return list(x) if isinstance(x, (list, tuple)) else [x]
        for _k in ("seeds", "durations_s", "loglike_mean"):
            meta_r[_k] = _as_list(meta_r.get(_k))

        C   = int(dims["chains"])
        S   = int(dims["n_samples"])
        K   = int(dims["K"])
        pc  = dims["per_chain"]
        has_Z = bool(dims["has_Z"])

        def stack(name):
            shape = pc[name]
            return np.stack(
                [_read_bin(raw_dir / f"{name}_chain{c}.bin", shape) for c in range(C)],
                axis=0,
            )

        mu_k   = stack("mu")                     # (C, S, K, P)
        prec   = stack("prec")                   # (C, S, K, P, P)
        pvec   = stack("pvec")                   # (C, S, K)
        beta_i = stack("beta")                   # (C, S, N, P)
        hlog(f"Loaded raw draws: chains={C}  draws/chain={S}  K={K}  P={P}")

        # ── 3. rebuild the Liesel-format latent + assemble canonical dict ──
        sigma_inv_chol_k_latent = _precision_to_latent(prec)   # (C, S, K, P(P+1)/2)
        if not np.all(np.isfinite(sigma_inv_chol_k_latent)):
            raise FloatingPointError(
                "Non-finite values in reconstructed sigma_inv_chol_k_latent "
                "(precision -> Cholesky -> FillScaleTriL inverse) - likely a "
                "degenerate/empty mixture component in the bayesm draws."
            )

        canonical = {
            "mu_k": mu_k,
            "sigma_inv_chol_k_latent": sigma_inv_chol_k_latent,
            "pvec": pvec,
            "beta_i": beta_i,
        }
        if has_Z:
            canonical["Delta"] = stack("delta")  # (C, S, D, P)

        with open(outdir / "posterior_raw.pkl", "wb") as f:
            pickle.dump({k: np.asarray(v) for k, v in canonical.items()}, f)
        hlog("Saved posterior_raw.pkl")

        # ── 4. export.pkl via the SHARED exporter (mu/sigma/std/pvec) ──────
        try:
            from src.analysis import export_posterior_to_pickle
            export_posterior_to_pickle(
                samples=canonical, K=args.k_model, P=P,
                filename="export.pkl", output_dir=str(outdir),
            )
        except Exception as exp_err:
            hlog(f"WARNING export failed: {exp_err}")

        duration = time.time() - t0

        # ── 5. meta.json - same keys the notebook reads (warmup<-burn_in,
        #        posterior<-kept draws), plus bayesm-specific fields ─────────
        meta = {
            "scenario": args.scenario, "data_path": str(data_path),
            "k_model": args.k_model, "k_true": K_TRUE,
            "n_params": P, "n_units": N, "n_demos": D,
            "sampler": "bayesm", "chains": args.chains,
            "warmup": args.burn_in, "posterior": S,        # mapped for notebook display
            "seed": args.seed,
            "a_delta": args.a_delta, "a_mu": args.a_mu, "dirichlet_a": args.dirichlet_a,
            "r_total": args.r_total, "thin": args.thin, "burn_in": args.burn_in,
            "n_samples": S, "seeds": meta_r.get("seeds"),
            "bayesm_version": meta_r.get("bayesm_version"),
            "r_version": meta_r.get("r_version"),
            "chain_durations_s": meta_r.get("durations_s"),
            "loglike_mean": meta_r.get("loglike_mean"),
            "started_at": started_at, "duration_s": round(duration, 1),
        }
        with open(outdir / "meta.json", "w") as f:
            json.dump(meta, f, indent=2)

        # ── 6. summary headline + per-chain timings/loglike ────────────────
        hlog("-" * 60)
        hlog("Per-chain summary (from R):")
        seeds = meta_r.get("seeds") or []
        durs  = meta_r.get("durations_s") or []
        lls   = meta_r.get("loglike_mean") or []
        for c in range(C):
            sd = seeds[c] if c < len(seeds) else "?"
            du = durs[c] if c < len(durs) else "?"
            ll = lls[c] if c < len(lls) else "?"
            hlog(f"   chain {c}  seed={sd:<6}  {du:>7}s  mean loglike={ll}")
        hlog("-" * 60)
        hlog(f"DONE in {datetime.timedelta(seconds=int(duration))}")
        summary_path.write_text("\n".join(summary_lines) + "\n", encoding="utf-8")

        # ── 7. tidy up the big raw bins (redundant with posterior_raw.pkl) ─
        if not args.keep_raw:
            for b in raw_dir.glob("*.bin"):
                b.unlink()

        with open(status_path, "w") as f:
            json.dump({"status": "success", "duration_s": round(duration, 1),
                       "finished_at": datetime.datetime.now().isoformat(timespec="seconds")},
                      f, indent=2)
        return 0

    except Exception:
        duration = time.time() - t0
        err = traceback.format_exc()
        hlog(err)
        try:
            summary_path.write_text("\n".join(summary_lines) + "\n", encoding="utf-8")
        except Exception:
            pass
        with open(status_path, "w") as f:
            json.dump({"status": "failed", "error": err, "duration_s": round(duration, 1),
                       "finished_at": datetime.datetime.now().isoformat(timespec="seconds")},
                      f, indent=2)
        return 1


if __name__ == "__main__":
    sys.exit(main())
